### DISCUSSION TITLE
Should README.md avoid encourage the use of mutable global attributes?

### DISCUSSION THREAD

**Discussion by sodul** [pydantic/pydantic#9565] [Question]
https://pythonforthelab.com/blog/mutable-and-immutable-attributes-of-classes/

I the README.md the example is showing the use of a class attribute that defaults to an empty list:
https://github.com/pydantic/pydantic/blob/main/README.md?plain=1#L51

This is usually considered dangerous because many less experienced python developers do not understand that when declaring a class attribute such as `friends: List[int] = []`, the empty list declaration is only interpreted once at import time, and that same list instance is now re-used as the default value for every instance of the class.

After spending some time searching more details about this I did find this conversation:
https://github.com/pydantic/pydantic/discussions/3274

Which lead me to the documentation explaining the actual behavior of pydantic:
https://docs.pydantic.dev/latest/concepts/models/#fields-with-non-hashable-default-values

Even if pydantic does replace the default value with a new instance, my concern is that less experienced Python developers will use the pattern for non pydantic objects. This leads to potential runtime bugs, that can be hard to debug for folks that are less familiar with this behavior of python.

The dataclasses from the standard library actually raises a ValueError unless a default factory is explicitly used: https://docs.python.org/3/library/dataclasses.html#mutable-default-values

I feel this breaks the second principle of the Zen of Python: "Explicit is better than implicit". I could also mention "[Special cases](https://en.wikipedia.org/wiki/Edge_case) aren't special enough to break the rules.", "There should be one-- and preferably only one --obvious way to do it".

I have debugged enough legacy code where the use of mutable defaults have been causing unexpected behavior, that I really feel that the pattern should be discouraged rather than encouraged and I would even be in favor of getting it deprecated and have it be consisted with the data classes of the standard library by raising `ValueError` in a future release of pydantic.

**Comment by samuelcolvin** [comment]
> There should be one-- and preferably only one --obvious way to do it
 
* Dataclasses
* Typeddict
* NamedTuple

I think that rule is well and truly dead.

---

More generally, I think our behavior is fine, and I think we should avoid "discouraging" things that work fine. In the end we're library authors, not teachers.

That said, I'd be happy to add a comment everywhere we do this in the docs explaining why it's okay, or linking to somewhere where we explain what we do.

**Reply by sodul** [reply]
Definitely good to clarify more, the documentation was rather hard to find. I had flagged a lot of these default lists to be fixed in our codebase as potential bugs, until I found out that it was actually 'ok'.
